example/common/flat_write_stream.hpp

```diff
@@ -0,0 +1,554 @@
+                    if(iter != buffers.end()) {
+                        write_some_op<
+                                ConstBufferSequence,
+                                boost::beast::handler_type<
+                                        WriteHandler, void(boost::beast::error_code, std::size_t)>> op(handler, *this, boost::asio::const_buffers_1{*iter}, std::shared_ptr<char>(nullptr));
+
+                        op(boost::beast::error_code{}, 0);
+                        return init.result.get();
+                    }
+                    else {
+                        if(size <= max_stack_size) {
+                            write_some_op<
+                                    ConstBufferSequence,
+                                    boost::beast::handler_type<
+                                            WriteHandler, void(boost::beast::error_code, std::size_t)>> op(handler, *this, stack_flatten(buffers), std::shared_ptr<char>(nullptr));
+
+                            op(boost::beast::error_code{}, 0);
+                            return init.result.get();
+                        } else {
+                            try {
+                                auto buf = std::shared_ptr<char>(new char[size], std::default_delete<char[]>());
+                                auto copy_size = boost::asio::buffer_copy(boost::asio::buffer(buf.get(), size),
+                                                                          buffers);
+                                write_some_op<
+                                        ConstBufferSequence,
+                                        boost::beast::handler_type<
+                                                WriteHandler, void(boost::beast::error_code, std::size_t)>> op(handler, *this, boost::asio::const_buffers_1(buf.get(), copy_size), buf);
+
+                                op(boost::beast::error_code{}, 0);
+                                return init.result.get();
+                            }
+                            catch (std::bad_alloc const &) {
+                                // todo
+                                return;
+                            }
+                        }
+                    }
+                }
+                else {
+                    return next_layer_.async_write_some(buffers, handler);
+                }
+            }
+
+            friend
+            void
+            teardown(websocket::role_type,
+                     flat_write_stream&, boost::system::error_code& ec)
+            {
+                ec.assign(0, ec.category());
+            }
+
+            template<class TeardownHandler>
+            friend
+            void
+            async_teardown(websocket::role_type,
+                           flat_write_stream& s, TeardownHandler&& handler)
+            {
+                std::cout << "async_teardown" << std::endl;   //todo: remove
+
+                s.get_io_service().post(
```

Comment:
Just call `async_teardown(stream.next_layer(), ...)` as with the synchronous case. Add the `using` statement so that argument dependent lookup will work.